### ISSUE TITLE
`at-rule-semicolon-space-before` reports the whitespace in front of a semicolon the file does not spell

### ISSUE THREAD

**Issue by firefoxic** [stylelint-stylistic/stylelint-stylistic#395]
`at-rule-semicolon-space-before` measures `rawNodeString(atRule)` and asks the whitespace checker about the position one character past its end, as though a semicolon always stood there:

```js
let nodeString = rawNodeString(atRule)
let problemIndex = nodeString.length - 1

checker.before({ source: nodeString, index: nodeString.length, … })
```

A bodiless at-rule that closes a block carries no semicolon, and `rawNodeString` hands back the node together with the whitespace on either side of it, the parent's own `raws.after` among it. So the last characters of that string are the whitespace in front of the closing brace, the rule reads them as the whitespace in front of a semicolon, and reports them. Plain CSS, no comment and no custom syntax:

```css
a { @import "x" }
```

`never` — `1:17 Unexpected whitespace before ";"`. Written over three lines, where the whitespace in front of the brace is a line break:

```css
a {
	@import "x"
}
```

`never` — `3:2 Unexpected whitespace before ";"`; `always` — `3:2 Expected single space before ";"`. The at-rule stands on line 2 in both.

At the top level of a file, where `rawNodeString` has no trailing whitespace to hand back, the same happens under `always` alone:

```css
@import "x"
```

`always` — `1:11 Expected single space before ";"`.

Adding the semicolon settles the two `never` warnings, which is what makes those look sound. It settles neither `always` one:

```css
@import "x"     /* always — 1:11 Expected single space before ";" */
@import "x";    /* always — 1:11 Expected single space before ";", the same warning */
```

and inside a block adding it moves the warning rather than removing it, from the closing brace on line 3 to the at-rule on line 2. So under `never` the rule asks in effect for a semicolon to be added, and under `always` it asks for a space in front of a character that is not there and goes on asking once it is.

## What that costs today

The rule writes no fix, so nothing in the file is corrupted — the cost is a warning about a character the file does not hold, which no writing of that file can be right about. `at-rule-semicolon-newline-after` reads the same node through the same guard and stays quiet on all three inputs above, so the two rules of the pair disagree over whether the semicolon is there.

## Why no oracle sees it

All six ask about the output of a fix or about a file's twin, and this rule writes nothing.

Found while measuring the reachability of the branch for #357, which is where the rule was run over every spelling of a bodiless at-rule.

**Comment by firefoxic** [comment]
One more shape of the same reading, met while reviewing the branch of #477: over an inline `style` attribute, `<a style="@foo bar">` under `postcss-html`, this rule reports `1:18 Expected single space before ";"` where no semicolon stands at all — the root of such an attribute carries no closing brace, so what the rule reads as the whitespace in front of a semicolon is the end of the attribute's text. `declaration-block-trailing-semicolon` leaves an at-rule of such a root alone by design, so nothing ever writes a semicolon there, and the warning has no fix to wait for. Same on `main` at `7c685af` and on that branch.
